File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsnonassociationsapi/util/TestData.kt
package uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.util

import uk.gov.justice.digital.hmpps.SYSTEM_USERNAME
import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.dto.CreateNonAssociationRequest
import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.dto.Reason
import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.dto.RestrictionType
import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.dto.Role
import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.dto.offendersearch.OffenderSearchPrisoner
import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.helper.TestBase
import uk.gov.justice.digital.hmpps.hmppsnonassociationsapi.jpa.NonAssociation
import java.time.Clock
import java.time.LocalDateTime

/**
 * Creates a new non-association entity
 */

fun genNonAssociation(
  id: Long? = null,
  firstPrisonerNumber: String,
  secondPrisonerNumber: String,
  createTime: LocalDateTime = LocalDateTime.now(TestBase.clock).minusDays(1),
  closed: Boolean = false,
  closedReason: String? = "No closure reason provided",
  updatedBy: String? = SYSTEM_USERNAME,
  clock: Clock = TestBase.clock,
  authorisedBy: String? = null,
) = NonAssociation(
  id = id,
  firstPrisonerNumber = firstPrisonerNumber,
  firstPrisonerRole = Role.PERPETRATOR,
  secondPrisonerNumber = secondPrisonerNumber,
  secondPrisonerRole = Role.VICTIM,
  comment = "Comment",
  reason = Reason.BULLYING,
  restrictionType = RestrictionType.WING,
  authorisedBy = authorisedBy,
  whenUpdated = createTime,
  whenCreated = createTime,
  updatedBy = updatedBy ?: "A_DPS_USER",
  isClosed = closed,
  closedAt = if (closed) {
    LocalDateTime.now(clock)
  } else {
    null
  },
  closedBy = if (closed) {
    "A_DPS_USER"
  } else {
    null
  },
  closedReason = if (closed) {
    closedReason
  } else {
    null
  },
)

/**
 * Returns a test CreateNonAssociationRequest
 *
 * Arguments are all optional with good (test) defaults so that you don't have to
 * pass all non-association fields when you just need a valid value.
 */
fun createNonAssociationRequest(
  firstPrisonerNumber: String = "A1234BC",
  firstPrisonerRole: Role = Role.VICTIM,
  secondPrisonerNumber: String = "D5678EF",
  secondPrisonerRole: Role = Role.PERPETRATOR,
  comment: String = "test comment",
  reason: Reason = Reason.THREAT,
  restrictionType: RestrictionType = RestrictionType.CELL,
): CreateNonAssociationRequest {
  return CreateNonAssociationRequest(
    firstPrisonerNumber = firstPrisonerNumber,
    firstPrisonerRole = firstPrisonerRole,
    secondPrisonerNumber = secondPrisonerNumber,
    secondPrisonerRole = secondPrisonerRole,
    comment = comment,
    reason = reason,
    restrictionType = restrictionType,
  )
}

val offenderSearchPrisoners = mapOf(
  "A1234BC" to OffenderSearchPrisoner(
    prisonerNumber = "A1234BC",
    firstName = "John",
    lastName = "Doe",
    prisonId = "MDI",
    prisonName = "Moorland",
    cellLocation = "MDI-A-1",
  ),
  "D5678EF" to OffenderSearchPrisoner(
    prisonerNumber = "D5678EF",
    firstName = "Merlin",
    lastName = "Somerplumbs",
    prisonId = "MDI",
    prisonName = "Moorland",
    cellLocation = "MDI-A-2",
  ),
  "G9012HI" to OffenderSearchPrisoner(
    prisonerNumber = "G9012HI",
    firstName = "Josh",
    lastName = "Plimburkson",
    prisonId = "MDI",
    prisonName = "Moorland",
    cellLocation = "MDI-A-3",
  ),
  // Different prison
  "L3456MN" to OffenderSearchPrisoner(
    prisonerNumber = "L3456MN",
    firstName = "Edward",
    lastName = "Lillibluprs",
    prisonId = "FBI",
    prisonName = "Forest Bank",
    cellLocation = "FBI-C-2",
  ),
  // In transfer
  "C1234CC" to OffenderSearchPrisoner(
    prisonerNumber = "C1234CC",
    firstName = "MAX",
    lastName = "CLARKE",
    prisonId = "TRN",
    prisonName = "Transfer",
    cellLocation = null,
  ),
  // Outside any establishment
  "B1234BB" to OffenderSearchPrisoner(
    prisonerNumber = "B1234BB",
    firstName = "JOE",
    lastName = "PETERS",
    prisonId = "OUT",
    prisonName = "Outside - released from Moorland (HMP)",
    cellLocation = null,
  ),
  // Null location, allegedly indicates no booking
  "D1234DD" to OffenderSearchPrisoner(
    prisonerNumber = "D1234DD",
    firstName = "NATHAN",
    lastName = "LOST",
    prisonId = null,
    prisonName = null,
    cellLocation = null,
  ),
)

File: clients/node/test/testData.ts
import type {
  OpenNonAssociation,
  ClosedNonAssociation,
  NonAssociationsList,
  OpenNonAssociationsListItem,
  ClosedNonAssociationsListItem,
} from '../src'

export const openNonAssociation: OpenNonAssociation = {
  id: 101,
  firstPrisonerNumber: 'A1234BC',
  firstPrisonerRole: 'PERPETRATOR',
  firstPrisonerRoleDescription: 'Perpetrator',
  secondPrisonerNumber: 'A1235EF',
  secondPrisonerRole: 'VICTIM',
  secondPrisonerRoleDescription: 'Victim',
  reason: 'THREAT',
  reasonDescription: 'Threat',
  restrictionType: 'CELL',
  restrictionTypeDescription: 'Cell only',
  comment: 'See IR 12133100',
  authorisedBy: 'cde87s',
  updatedBy: 'cde87s',
  whenCreated: new Date('2023-07-21T08:14:21'),
  whenUpdated: new Date('2023-07-21T08:14:21'),
  isClosed: false,
  closedBy: null,
  closedReason: null,
  closedAt: null,
}

export const closedNonAssociation: ClosedNonAssociation = {
  ...openNonAssociation,
  isClosed: true,
  closedBy: 'abc12a',
  closedReason: 'Problem solved',
  closedAt: new Date('2023-07-26T12:34:56'),
}

export const nonAssociationListOpen: NonAssociationsList<OpenNonAssociationsListItem> = {
  prisonId: 'MDI',
  prisonName: 'Moorland (HMP & YOI)',
  prisonerNumber: 'A1234BC',
  firstName: 'DAVID',
  lastName: 'JONES',
  cellLocation: '1-1-001',
  openCount: 1,
  closedCount: 0,
  nonAssociations: [
    {
      id: 101,
      role: 'PERPETRATOR',
      roleDescription: 'Perpetrator',
      reason: 'THREAT',
      reasonDescription: 'Threat',
      restrictionType: 'CELL',
      restrictionTypeDescription: 'Cell only',
      comment: 'See IR 12133100',
      otherPrisonerDetails: {
        prisonId: 'MDI',
        prisonName: 'Moorland (HMP & YOI)',
        firstName: 'DAVID',
        lastName: 'JONES',
        prisonerNumber: 'A1235EF',
        role: 'VICTIM',
        roleDescription: 'Victim',
      },
      authorisedBy: 'cde87s',
      updatedBy: 'cde87s',
      whenCreated: new Date('2023-07-26T12:34:56'),
      whenUpdated: new Date('2023-07-26T12:34:56'),
      isClosed: false,
      closedBy: null,
      closedReason: null,
      closedAt: null,
    },
  ],
}

export const nonAssociationListClosed: NonAssociationsList<ClosedNonAssociationsListItem> = {
  ...nonAssociationListOpen,
  openCount: 0,
  closedCount: 1,
  nonAssociations: [
    {
      ...nonAssociationListOpen.nonAssociations[0],
      isClosed: true,
      closedBy: 'lev79n',
      closedReason: 'Problem solved',
      closedAt: new Date('2023-07-27T12:34:56'),
      whenUpdated: new Date('2023-07-27T12:34:56'),
    },
  ],
}
